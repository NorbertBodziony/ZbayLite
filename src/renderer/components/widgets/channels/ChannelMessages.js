import React from 'react'
import PropTypes from 'prop-types'
import Immutable from 'immutable'
import * as R from 'ramda'
import { Scrollbars } from 'react-custom-scrollbars'
import { withContentRect } from 'react-measure'

import Grid from '@material-ui/core/Grid'
import List from '@material-ui/core/List'
import RootRef from '@material-ui/core/RootRef'
import { withStyles } from '@material-ui/core/styles'

import ChannelMessage from './ChannelMessage'

const styles = theme => ({
  wrapper: {
    background: theme.palette.primary.light,
    flexGrow: 1
  },
  list: {
    padding: '0 8px',
    width: '100%'
  }
})

// TODO: scrollbar smart pagination
export const ChannelMessages = ({ classes, messages, measureRef, contentRect }) => {
  const scrollbarRef = ref => ref && ref.scrollToBottom()
  return (
    <RootRef rootRef={measureRef}>
      <Grid container direction='column' justify='flex-end' className={classes.wrapper} >
        <Scrollbars
          autoHide
          autoHeight
          autoHeightMax={contentRect.bounds.height}
          ref={scrollbarRef}
          autoHideTimeout={500}
        >
          <List disablePadding className={classes.list}>
            { messages.map(msg => (<ChannelMessage key={msg.get('id')} message={msg} />))}
          </List>
        </Scrollbars>
      </Grid>
    </RootRef>
  )
}

ChannelMessages.propTypes = {
  classes: PropTypes.object.isRequired,
  messages: PropTypes.instanceOf(Immutable.List).isRequired,
  measureRef: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({ current: PropTypes.instanceOf(React.Element) })
  ]).isRequired,
  contentRect: PropTypes.shape({
    bounds: PropTypes.shape({
      height: PropTypes.number
    }).isRequired
  })
}

ChannelMessages.defaultProps = {
  messages: []
}

export default R.compose(
  React.memo,
  withStyles(styles),
  withContentRect('bounds')
)(ChannelMessages)
