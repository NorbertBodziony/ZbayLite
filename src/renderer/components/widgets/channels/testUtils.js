export const createChannel = id => ({
  name: `Channel ${id}`,
  description: id % 2 === 0 ? '' : `Channel about ${id}`,
  private: id % 2 === 0,
  unread: id,
  hash: `test-hash-${id}`,
  address: `zs1testaddress${id}`
})
