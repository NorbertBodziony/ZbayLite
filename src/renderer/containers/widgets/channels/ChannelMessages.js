import React, { useEffect } from 'react'
import { connect } from 'react-redux'
import Immutable from 'immutable'

import ChannelMessagesComponent from '../../../components/widgets/channels/ChannelMessages'
import channelSelectors from '../../../store/selectors/channel'
import contactsSelectors from '../../../store/selectors/contacts'
import offersSelectors from '../../../store/selectors/offers'

export const mapStateToProps = (state, { contactId, signerPubKey }) => {
  const offersMessages = offersSelectors.offer(contactId)(state)
  return {
    messages: contactId
      ? offersMessages
        ? offersSelectors.offerMessages(contactId, signerPubKey)(state)
        : contactsSelectors.directMessages(contactId, signerPubKey)(state)
      : channelSelectors.messages(signerPubKey)(state),
    channelId: channelSelectors.channelId(state)
  }
}

export const ChannelMessages = ({ className, messages, contactId, channelId, contentRect }) => {
  const [scrollPosition, setScrollPosition] = React.useState(-1)
  useEffect(
    () => {
      setScrollPosition(-1)
    },
    [channelId, contactId]
  )
  return (
    <ChannelMessagesComponent
      scrollPosition={scrollPosition}
      setScrollPosition={setScrollPosition}
      messages={messages}
      contactId={contactId}
      contentRect={contentRect}
    />
  )
}

export default connect(mapStateToProps)(
  React.memo(ChannelMessages, (before, after) => {
    return Immutable.is(before.messages, after.messages)
  })
)
