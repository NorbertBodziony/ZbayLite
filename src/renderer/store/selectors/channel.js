import { createSelector } from 'reselect'
import channelsSelector from './channels'
import identitySelectors from './identity'
import pendingMessagesSelectors from './pendingMessages'

const store = s => s

const channel = createSelector(store, state => state.get('channel'))

const data = createSelector(
  channelsSelector.data,
  channel,
  (channels, channel) => channels.find(ch => ch.get('id') === channel.id)
)

export const spentFilterValue = createSelector(channel, c => c.get('spentFilterValue', -1))

export const message = createSelector(channel, c => c.get('message'))

export const pendingMessages = createSelector(
  pendingMessagesSelectors.pendingMessages,
  channel,
  (pendingMessages, channel) => pendingMessages.filter(m => m.channelId === channel.id)
)

export const messagesMeta = createSelector(channel, c => c.messages)

const _isOwner = (address, message) => message.getIn(['sender', 'replyTo']) === address

export const messages = createSelector(
  identitySelectors.data,
  messagesMeta,
  pendingMessages,
  (identity, meta, pendingMessages) => {
    const displayableBroadcasted = meta.data.map(
      m => m.set('fromYou', _isOwner(identity.address, m))
    )
    const displayablePending = pendingMessages.map(
      m => m.get('message')
        .set('error', m.get('error'))
        .set('status', m.get('status'))
        .set('id', m.get('opId'))
        .set('fromYou', _isOwner(identity.address, m.get('message')))
    )
    return displayableBroadcasted.concat(displayablePending.values())
      .sortBy(m => m.get('createdAt'))
  })

export default {
  data,
  channel,
  spentFilterValue,
  message,
  pendingMessages,
  messages
}
