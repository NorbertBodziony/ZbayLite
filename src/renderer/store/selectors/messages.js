import { createSelector } from 'reselect'
import Immutable from 'immutable'

const store = s => s

const messages = createSelector(
  store,
  state => state.get('messages')
)
const currentChannelMessages = channelId =>
  createSelector(
    messages,
    msgs => msgs.getIn([channelId, 'messages'], Immutable.List())
  )
const allMessages = createSelector(
  messages,
  msgs =>
    Immutable.List(msgs.keySeq()).reduce((acc, channelId) => {
      return acc.concat(msgs.getIn([channelId, 'messages']))
    }, Immutable.List())
)
const messageById = id =>
  createSelector(
    allMessages,
    msgs => {
      return msgs.find(message => message.id === id)
    }
  )

const channelSettingsMessages = channelId => createSelector(
  messages,
  msgs => msgs.getIn([channelId, 'messages'], Immutable.List()).filter(msg => msg.get('type') === 6)
)

const channelOwner = channelId => createSelector(
  channelSettingsMessages(channelId),
  msgs => {
    let channelOwner = null
    channelOwner = msgs.get(0) ? msgs.get(0).get('publicKey') : null
    for (const msg of msgs) {
      if (channelOwner === msg.get('publicKey')) {
        channelOwner = msg.getIn(['message', 'owner'])
      }
    }
    return channelOwner
  }
)

export default {
  messages,
  channelOwner,
  channelSettingsMessages,
  currentChannelMessages,
  messageById,
  allMessages
}
