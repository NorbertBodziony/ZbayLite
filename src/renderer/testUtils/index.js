import { DateTime, Settings } from 'luxon'
import BigNumber from 'bignumber.js'

import { messages as zbayMessages } from '../zbay'

Settings.defaultZoneName = 'utc'

export const now = DateTime.utc(2019, 3, 7, 13, 3, 48)

export const createChannel = id => ({
  id: id,
  name: `Channel ${id}`,
  private: Boolean(id % 2),
  address: `zs1testaddress${id}`,
  unread: id,
  description: id % 2 === 0 ? '' : `Channel about ${id}`,
  keys: {
    ivk: `incoming-viewing-key-${id}`
  }
})

export const channels = {
  createChannel
}

export const createMessage = (
  id,
  createdAt = now.minus({ hours: id }).toSeconds(),
  replyTo = 'zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya'
) => ({
  id,
  type: zbayMessages.messageType.BASIC,
  createdAt,
  message: `This is some message ${id}`,
  sender: {
    replyTo
  }
})

export const createReceivedMessage = ({
  id,
  createdAt = now.toSeconds()
}) => ({
  id,
  spent: new BigNumber('0.32'),
  type: zbayMessages.messageType.BASIC,
  createdAt,
  message: `This is a message with id ${id}`,
  sender: {
    replyTo: 'zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya',
    username: 'Wenus'
  }
})

export const createSendableMessage = ({
  message,
  createdAt = now.toSeconds()
}) => ({
  type: zbayMessages.messageType.BASIC,
  sender: {
    replyTo: 'zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya',
    username: 'Wenus'
  },
  createdAt,
  message
})

export const createSendableTransferMessage = ({
  message,
  createdAt = now.toSeconds()
}) => ({
  type: zbayMessages.messageType.BASIC,
  sender: {
    replyTo: 'zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya',
    username: 'Wenus'
  },
  receiver: {
    replyTo: 'zs1z9rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya',
    username: 'Sun'
  },
  createdAt,
  message
})

export const createReceivedTransferMessage = ({ id, createdAt = now.toSeconds() }) => ({
  id,
  spent: new BigNumber('0.32'),
  type: zbayMessages.messageType.TRANSFER,
  createdAt,
  message: `This is a message with id ${id}`,
  sender: {
    replyTo: 'zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya',
    username: 'Wenus'
  },
  receiver: {
    replyTo: 'zs1z9rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya',
    username: 'Sun'
  }
})

export const messages = {
  createReceivedMessage,
  createSendableMessage,
  createReceivedTransferMessage,
  createSendableTransferMessage,
  createMessage
}

export const createTransfer = ({
  txid,
  amount = '0.0001',
  memo = '',
  change = false
}) => ({
  txid,
  amount,
  memo,
  change
})

export const transfers = {
  createTransfer
}

export const createIdentity = ({
  name = 'Saturn',
  address = 'saturn-private-address',
  transparentAddress = 'saturn-transparent-address',
  keys = {
    tpk: 'saturn-tpk',
    sk: 'saturn-sk'
  }
} = {}) => ({
  name,
  address,
  transparentAddress,
  keys
})

export default {
  now,
  transfers,
  channels,
  messages,
  createIdentity
}
