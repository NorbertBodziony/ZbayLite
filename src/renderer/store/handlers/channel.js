import Immutable from 'immutable'
import BigNumber from 'bignumber.js'
import * as R from 'ramda'
import { createAction, handleActions } from 'redux-actions'

import operationsHandlers, { operationTypes, PendingMessageOp } from './operations'
import notificationsHandlers from './notifications'
import channelSelectors from '../selectors/channel'
import identitySelectors from '../selectors/identity'
import nodeSelectors from '../selectors/node'
import operationsSelectors from '../selectors/operations'
import { getClient } from '../../zcash'
import { messages } from '../../zbay'
import { errorNotification, LoaderState } from './utils'
import { channelToUri } from '../../zbay/channels'

export const MessagesState = Immutable.Record({
  loader: LoaderState({ loading: true }),
  data: Immutable.List(),
  errors: ''
})

export const ChannelState = Immutable.Record({
  spentFilterValue: new BigNumber(0),
  id: null,
  message: '',
  shareableUri: '',
  messages: MessagesState(),
  members: null
}, 'ChannelState')

export const initialState = ChannelState()

const setSpentFilterValue = createAction('SET_SPENT_FILTER_VALUE', (_, value) => value)
const setMessage = createAction('SET_CHANNEL_MESSAGE', R.path(['target', 'value']))
const setChannelId = createAction('SET_CHANNEL_ID')
const setMessages = createAction('SET_CHANNEL_MESSAGES')
const setLoading = createAction('SET_CHANNEL_LOADING')
const setLoadingMessage = createAction('SET_CHANNEL_LOADING_MESSAGE')
const setShareableUri = createAction('SET_CHANNEL_SHAREABLE_URI')

export const actions = {
  setLoading,
  setLoadingMessage,
  setSpentFilterValue,
  setMessage,
  setShareableUri,
  setChannelId
}

const loadMessages = () => async (dispatch, getState) => {
  const channel = channelSelectors.data(getState())
  const isTestnet = nodeSelectors.node(getState()).isTestnet
  const transfers = await getClient().payment.received(channel.get('address'))

  const pendingMessages = operationsSelectors.pendingMessages(getState())

  const msgs = await Promise.all(transfers.map(
    async (transfer) => {
      const message = await messages.transferToMessage(transfer, isTestnet)

      const pendingMessage = pendingMessages.find(
        pm => pm.txId && pm.txId === message.id)
      if (pendingMessage) {
        dispatch(
          operationsHandlers.actions.removeOperation(pendingMessage.opId)
        )
      }
      return message
    }
  ))

  dispatch(setMessages(msgs))
}

const loadChannel = (id) => async (dispatch, getState) => {
  dispatch(setLoading(true))
  dispatch(setLoadingMessage('Loading messages'))
  try {
    dispatch(setChannelId(id))

    // Calculate URI on load, that way it won't be outdated, even if someone decides
    // to update channel in vault manually
    const channel = channelSelectors.data(getState())
    const uri = await channelToUri(channel.toJS())
    dispatch(setShareableUri(uri))

    await dispatch(loadMessages())
  } catch (err) {}
  dispatch(setLoading(false))
}

const sendOnEnter = (event) => async (dispatch, getState) => {
  const enterPressed = event.nativeEvent.keyCode === 13
  const shiftPressed = event.nativeEvent.shiftKey === true
  if (enterPressed && !shiftPressed) {
    event.preventDefault()
    const message = messages.createMessage({
      identity: identitySelectors.data(getState()).toJS(),
      messageData: {
        type: messages.messageType.BASIC,
        data: event.target.value
      }
    })
    const channel = channelSelectors.data(getState()).toJS()
    const transfer = await messages.messageToTransfer({
      message,
      channel
    })
    try {
      const opId = await getClient().payment.send(transfer)
      await dispatch(operationsHandlers.epics.observeOperation({
        opId,
        type: operationTypes.pendingMessage,
        meta: PendingMessageOp({
          channelId: channel.id,
          message: Immutable.fromJS(message)
        })
      }))
    } catch (err) {
      notificationsHandlers.actions.enqueueSnackbar(
        errorNotification({
          message: 'Couldn\'t send the message, please check node connection.'
        })
      )
    }
    dispatch(setMessage(''))
  }
}

const resendMessage = (message) => async (dispatch, getState) => {
  dispatch(operationsHandlers.actions.removeOperation(message.id))
  const channel = channelSelectors.data(getState()).toJS()
  const transfer = await messages.messageToTransfer({
    message,
    channel
  })
  try {
    const opId = await getClient().payment.send(transfer)
    await dispatch(operationsHandlers.epics.observeOperation({
      opId,
      type: operationTypes.pendingMessage,
      meta: PendingMessageOp({
        channelId: channel.id,
        message: Immutable.fromJS(message)
      })
    }))
  } catch (err) {
    notificationsHandlers.actions.enqueueSnackbar(
      errorNotification({
        message: 'Couldn\'t send the message, please check node connection.'
      })
    )
  }
}

export const epics = {
  sendOnEnter,
  loadChannel,
  resendMessage,
  loadMessages
}

// TODO: we should have a global loader map
export const reducer = handleActions({
  [setLoading]: (state, { payload: loading }) => state.setIn(['messages', 'loader', 'loading'], loading),
  [setLoadingMessage]: (state, { payload: message }) => state.setIn(['messages', 'loader', 'message'], message),
  [setSpentFilterValue]: (state, { payload: value }) => state.set('spentFilterValue', new BigNumber(value)),
  [setMessage]: (state, { payload: value }) => state.set('message', value),
  [setChannelId]: (state, { payload: id }) => state.set('id', id),
  [setShareableUri]: (state, { payload: uri }) => state.set('shareableUri', uri),
  [setMessages]: (state, { payload: messages }) => state.update(
    'messages',
    msgMeta => msgMeta
      .set('data', Immutable.fromJS(messages))
  )
}, initialState)

export default {
  reducer,
  epics,
  actions
}
