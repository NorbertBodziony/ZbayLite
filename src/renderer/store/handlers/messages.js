import Immutable from 'immutable'
import BigNumber from 'bignumber.js'
import * as R from 'ramda'
import { createAction, handleActions } from 'redux-actions'

import selectors from '../selectors/messages'
import appSelectors from '../selectors/app'
import channelsSelectors from '../selectors/channels'
import usersSelectors from '../selectors/users'
import identitySelectors from '../selectors/identity'
import operationsSelectors from '../selectors/operations'
import operationsHandlers from './operations'
import channelsHandlers from './channels'
import txnTimestampsHandlers from '../handlers/txnTimestamps'
import txnTimestampsSelector from '../selectors/txnTimestamps'
import appHandlers from './app'
import { messageType } from '../../zbay/messages'
import { messages as zbayMessages } from '../../zbay'
import { getClient } from '../../zcash'
import { displayMessageNotification } from '../../notifications'
import { getVault } from '../../vault'

export const MessageSender = Immutable.Record(
  {
    replyTo: '',
    username: ''
  },
  'MessageSender'
)

const _ReceivedMessage = Immutable.Record(
  {
    id: null,
    type: messageType.BASIC,
    sender: MessageSender(),
    createdAt: 0,
    message: '',
    spent: new BigNumber(0),
    isUnregistered: false,
    tag: '',
    offerOwner: ''
  },
  'ReceivedMessage'
)

export const ReceivedMessage = values => {
  const record = _ReceivedMessage(values)
  return record.set('sender', MessageSender(record.sender))
}

export const ChannelMessages = Immutable.Record(
  {
    messages: Immutable.List(),
    newMessages: Immutable.List()
  },
  'ChannelMessages'
)

export const initialState = Immutable.Map()

const setMessages = createAction('SET_MESSAGES')
const cleanNewMessages = createAction('CLEAN_NEW_MESSAGESS')
const appendNewMessages = createAction('APPEND_NEW_MESSAGES')

export const actions = {
  setMessages,
  cleanNewMessages,
  appendNewMessages
}

export const fetchMessages = channel => async (dispatch, getState) => {
  try {
    const pendingMessages = operationsSelectors.pendingMessages(getState())
    const identityAddress = identitySelectors.address(getState())
    const users = usersSelectors.users(getState())
    const txnTimestamps = txnTimestampsSelector.tnxTimestamps(getState())

    if (pendingMessages.find(msg => msg.status === 'pending')) {
      return
    }
    const channelId = channel.get('id')
    const previousMessages = selectors.currentChannelMessages(channelId)(getState())

    const transfers = await getClient().payment.received(channel.get('address'))

    if (transfers.length === appSelectors.transfers(getState()).get(channelId)) {
      return
    } else {
      dispatch(appHandlers.actions.setTransfers({ id: channelId, value: transfers.length }))
    }
    await transfers.forEach(async transfer => {
      if (!txnTimestamps.get(transfer.txid)) {
        const result = await getClient().confirmations.getResult(transfer.txid)
        await getVault().transactionsTimestamps.addTransaction(transfer.txid, result.blocktime)
        await dispatch(
          txnTimestampsHandlers.actions.addTxnTimestamp({
            tnxs: { [transfer.txid]: result.blocktime.toString() }
          })
        )
      }
    })
    const sortedTransfers = transfers.sort(
      (a, b) => txnTimestamps.get(a.txid) - txnTimestamps.get(b.txid)
    )
    const messagesAll = await Promise.all(
      sortedTransfers.map(async transfer => {
        const message = await zbayMessages.transferToMessage(transfer, users)
        if (message === null) {
          return ReceivedMessage(message)
        }
        const pendingMessage = pendingMessages.find(pm => pm.txId && pm.txId === message.id)
        if (pendingMessage) {
          dispatch(operationsHandlers.actions.removeOperation(pendingMessage.opId))
        }
        return ReceivedMessage(message)
      })
    )
    const messages = messagesAll.filter(message => message.id !== null)
    let lastSeen = channelsSelectors.lastSeen(channelId)(getState())
    if (!lastSeen) {
      await dispatch(channelsHandlers.epics.updateLastSeen({ channelId }))
      lastSeen = channelsSelectors.lastSeen(channelId)(getState())
    }
    const newMessages = zbayMessages.calculateDiff({
      previousMessages,
      nextMessages: Immutable.List(messages),
      lastSeen,
      identityAddress
    })
    await dispatch(channelsHandlers.actions.setUnread({ channelId, unread: newMessages.size }))
    await dispatch(
      appendNewMessages({
        channelId,
        messagesIds: newMessages.map(R.prop('id'))
      })
    )
    await dispatch(setMessages({ messages, channelId }))
    newMessages.map(nm => displayMessageNotification({ message: nm, channel }))
  } catch (err) {
    console.warn(err)
  }
}

export const epics = {
  fetchMessages
}

export const reducer = handleActions(
  {
    [setMessages]: (state, { payload: { channelId, messages } }) =>
      state.update(channelId, ChannelMessages(), cm =>
        cm.set('messages', Immutable.fromJS(messages))
      ),
    [cleanNewMessages]: (state, { payload: { channelId } }) =>
      state.update(channelId, ChannelMessages(), cm => cm.set('newMessages', Immutable.List())),
    [appendNewMessages]: (state, { payload: { channelId, messagesIds } }) =>
      state.update(channelId, ChannelMessages(), cm =>
        cm.update('newMessages', nm => nm.concat(messagesIds))
      )
  },
  initialState
)

export default {
  reducer,
  epics,
  actions
}
