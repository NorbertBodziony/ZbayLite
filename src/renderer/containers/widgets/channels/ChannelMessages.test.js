/* eslint import/first: 0 */
jest.mock('../../../vault')
import Immutable from 'immutable'
import BigNumber from 'bignumber.js'
import * as R from 'ramda'

import { mapStateToProps } from './ChannelMessages'

import { createReceivedMessage, now } from '../../../testUtils'
import create from '../../../store/create'
import { ChannelState } from '../../../store/handlers/channel'
import { ReceivedMessage, ChannelMessages } from '../../../store/handlers/messages'

describe('ChannelInput', () => {
  let store = null
  beforeEach(() => {
    jest.clearAllMocks()
    const channelId = 'this-is-test-channel-id'
    store = create({
      initialState: Immutable.Map({
        channel: ChannelState({
          spentFilterValue: 38,
          id: channelId,
          members: new BigNumber(0),
          message: 'This is a test message'
        }),
        messages: Immutable.Map({
          [channelId]: ChannelMessages({
            messages: Immutable.List(Immutable.fromJS(
              R.range(0, 4).map(id => ReceivedMessage(
                createReceivedMessage({ id, createdAt: now.minus({ hours: 2 * id }).toSeconds() })
              ))
            ))
          })
        })
      })
    })
  })

  it('will receive right props', async () => {
    const props = mapStateToProps(store.getState())
    expect(props).toMatchSnapshot()
  })
})
