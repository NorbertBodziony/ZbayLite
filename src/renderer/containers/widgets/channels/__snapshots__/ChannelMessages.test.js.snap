// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChannelInput will receive right props for channel 1`] = `
Object {
  "channelData": Immutable.Map {
    "keys": Immutable.Map {},
  },
  "channelId": "this-is-test-channel-id",
  "isInitialLoadFinished": false,
  "messages": Immutable.List [
    Immutable.DisplayableMessage {
      "id": 3,
      "type": 1,
      "sender": Immutable.ExchangeParticipant {
        "replyTo": "",
        "username": "",
      },
      "receiver": Immutable.ExchangeParticipant {
        "replyTo": "",
        "username": "Unnamed",
      },
      "createdAt": 1551942228,
      "message": "This is a message with id 3",
      "spent": "0.32",
      "fromYou": true,
      "status": "broadcasted",
      "error": null,
      "shippingData": null,
      "tag": "",
      "offerOwner": "",
      "isUnregistered": false,
      "publicKey": null,
      "blockTime": 9007199254740991,
    },
    Immutable.DisplayableMessage {
      "id": 2,
      "type": 1,
      "sender": Immutable.ExchangeParticipant {
        "replyTo": "",
        "username": "",
      },
      "receiver": Immutable.ExchangeParticipant {
        "replyTo": "",
        "username": "Unnamed",
      },
      "createdAt": 1551949428,
      "message": "This is a message with id 2",
      "spent": "0.32",
      "fromYou": true,
      "status": "broadcasted",
      "error": null,
      "shippingData": null,
      "tag": "",
      "offerOwner": "",
      "isUnregistered": false,
      "publicKey": null,
      "blockTime": 9007199254740991,
    },
    Immutable.DisplayableMessage {
      "id": 1,
      "type": 1,
      "sender": Immutable.ExchangeParticipant {
        "replyTo": "",
        "username": "",
      },
      "receiver": Immutable.ExchangeParticipant {
        "replyTo": "",
        "username": "Unnamed",
      },
      "createdAt": 1551956628,
      "message": "This is a message with id 1",
      "spent": "0.32",
      "fromYou": true,
      "status": "broadcasted",
      "error": null,
      "shippingData": null,
      "tag": "",
      "offerOwner": "",
      "isUnregistered": false,
      "publicKey": null,
      "blockTime": 9007199254740991,
    },
    Immutable.DisplayableMessage {
      "id": 0,
      "type": 1,
      "sender": Immutable.ExchangeParticipant {
        "replyTo": "",
        "username": "",
      },
      "receiver": Immutable.ExchangeParticipant {
        "replyTo": "",
        "username": "Unnamed",
      },
      "createdAt": 1551963828,
      "message": "This is a message with id 0",
      "spent": "0.32",
      "fromYou": true,
      "status": "broadcasted",
      "error": null,
      "shippingData": null,
      "tag": "",
      "offerOwner": "",
      "isUnregistered": false,
      "publicKey": null,
      "blockTime": 9007199254740991,
    },
  ],
  "network": "testnet",
  "publicChannels": Immutable.Map {},
  "qMessages": Immutable.Map {},
  "triggerScroll": false,
  "users": Immutable.Map {},
}
`;
